fulp_modules/LOOC_port/LOOC.dm

```diff
@@ -0,0 +1,88 @@
+//looc ported from citadel
+GLOBAL_VAR_INIT(LOOC_COLOR, null)//If this is null, use the CSS for OOC. Otherwise, use a custom colour.
+GLOBAL_VAR_INIT(normal_looc_colour, "#ec0303")
+
+/client/verb/looc(msg as text)
+	set name = "LOOC"
+	set desc = "Local OOC, seen only by those in view."
+	set category = "OOC"
+
+	if(GLOB.say_disabled)	//This is here to try to identify lag problems
+		to_chat(usr, "<span class='danger'> Speech is currently admin-disabled.</span>")
+		return
+
+	if(!mob)
+		return
+
+	msg = copytext_char(sanitize(msg), 1, MAX_MESSAGE_LEN)
+	if(!msg)
+		return
+
+	if(!(prefs.chat_toggles & CHAT_OOC))
+		to_chat(src, "<span class='danger'> You have OOC muted.</span>")
+		return
+
+	if(!holder)
+		if(!GLOB.ooc_allowed)
+			to_chat(src, "<span class='danger'> OOC is globally muted</span>")
+			return
+		if(prefs.muted & MUTE_OOC)
+			to_chat(src, "<span class='danger'> You cannot use OOC (muted).</span>")
+			return
+		if(handle_spam_prevention(msg,MUTE_OOC))
+			return
+		if(findtext(msg, "byond://"))
+			to_chat(src, "<B>Advertising other servers is not allowed.</B>")
+			log_admin("[key_name(src)] has attempted to advertise in LOOC: [msg]")
+			return
+		if(mob.stat)
+			to_chat(src, "<span class='danger'>You cannot use LOOC while unconscious or dead.</span>")
+			return
+		if(isdead(mob))
+			to_chat(src, "<span class='danger'>You cannot use LOOC while ghosting.</span>")
+			return
+
+
+
+	msg = emoji_parse(msg)
+
+	mob.log_talk(msg,LOG_OOC, tag="LOOC")
+
+	var/list/heard = get_hearers_in_view(7, get_top_level_mob(src.mob))
+	for(var/mob/M in heard)
+		if(!M.client)
+			continue
+		var/client/C = M.client
+		if (C in GLOB.admins)
+			continue //they are handled after that
+
+		if (isobserver(M))
+			continue //Also handled later.
+
+		if(C.prefs.chat_toggles & CHAT_OOC)
+			if(GLOB.LOOC_COLOR)
+				to_chat(C, "<font color='[GLOB.LOOC_COLOR]'><b><span class='prefix'>LOOC:</span> <EM>[src.mob.name]:</EM> <span class='message'>[msg]</span></b></font>")
+			else
+				to_chat(C, "<span class='looc'><span class='prefix'>LOOC:</span> <EM>[src.mob.name]:</EM> <span class='message'>[msg]</span></span>")
+
+	for(var/client/C in GLOB.admins)
+		if(C.prefs.chat_toggles & CHAT_OOC)
+			var/prefix = "(R)LOOC"
+			if (C.mob in heard)
+				prefix = "LOOC"
+			if(GLOB.LOOC_COLOR)
+				to_chat(C, "<font color='[GLOB.LOOC_COLOR]'><b>[ADMIN_FLW(usr)] <span class='prefix'>[prefix]:</span> <EM>[src.key]/[src.mob.name]:</EM> <span class='message'>[msg]</span></b></font>")
+			else
+				to_chat(C, "<span class='looc'>[ADMIN_FLW(usr)] <span class='prefix'>[prefix]:</span> <EM>[src.key]/[src.mob.name]:</EM> <span class='message'>[msg]</span></span>")
+
+/mob/proc/get_top_level_mob()
+    if(istype(src.loc,/mob)&&src.loc!=src)
+        var/mob/M=src.loc
+        return M.get_top_level_mob()
+    return src
+
+proc/get_top_level_mob(var/mob/S)
```

Comment:
[Copy-pasted](https://github.com/Citadel-Station-13/Citadel-Station-13/blob/9f50ea66c80485af8dc0975a7bf8b3f32ea41cf1/code/modules/mob/say_vr.dm#L11)

```suggestion
/proc/get_top_level_mob(mob/S)
```

Why is this necessary though?